Foundation/Parser.hs

```diff
@@ -276,3 +308,76 @@ skipWhile p = Parser $ \buf err ok ->
 -- stream
 skipAll :: Sequential input => Parser input ()
 skipAll = Parser $ \buf err ok -> runParser flushAll buf err ok
+
+data Count = Never | Once | Twice | Other Int
+  deriving (Show)
+instance Enum Count where
+    toEnum 0 = Never
+    toEnum 1 = Once
+    toEnum 2 = Twice
+    toEnum n
+        | n > 2 = Other n
+        | otherwise = Never
+    fromEnum Never = 0
+    fromEnum Once = 1
+    fromEnum Twice = 2
+    fromEnum (Other n) = n
+    succ Never = Once
+    succ Once = Twice
+    succ Twice = Other 3
+    succ (Other n)
+        | n == 0 = Once
+        | n == 1 = Twice
+        | otherwise = Other (succ n)
+    pred Never = Never
+    pred Once = Never
+    pred Twice = Once
+    pred (Other n)
+        | n == 2 = Once
+        | n == 3 = Twice
+        | otherwise = Other (pred n)
+
+data Condition = Exactly Count
+               | Between Count Count
+  deriving (Show)
+
+shouldStop :: Condition -> Bool
+shouldStop (Exactly   Never) = True
+shouldStop (Between _ Never) = True
+shouldStop _                 = False
+
+canStop :: Condition -> Bool
+canStop (Exactly Never)   = True
+canStop (Between Never _) = True
+canStop _                 = False
+
+decrement :: Condition -> Condition
+decrement (Exactly n)   = Exactly (pred n)
+decrement (Between a b) = Between (pred a) (pred b)
+
+-- | repeat the given Parser a given amount of time
+--
+-- If you know you want it to exactly perform a given amount of time:
+--
+-- ```
+-- repeat (Exactly Twice) (element 'a')
+-- ```
+--
+-- If you know your parser must performs from 0 to 8 times:
+--
+-- ```
+-- repeat (Between Never (Other 8))
+-- ```
+--
+-- *This interface is still WIP* but went handy when writting the IPv4/IPv6
+-- parsers.
+--
+repeat :: Sequential input => Condition -> Parser input a -> Parser input [a]
```

Comment:
This really looks like `replicateM` with a more flexible amount of repetition (variable). is it really worth it though.

for example in the parsing the ipv4 digits, where you expect `Bound Once (Other 3)`, you need to check after the fact that it's not > 255 (e.g. `127.0.0.258`), so you still need to check after parsing that the number fits. why not just `take1While isDigit` and check that it's valid just like you need to do anyway ?